MREUnityRuntime/MREUnityRuntimeLib/Core/SoundManager.cs

```diff
@@ -0,0 +1,142 @@
+        }
+
+        public void TrackUnpauseSound(Guid id)
+        {
+            _unpausedSoundInstances.Add(id);
+        }
+
+        public void ApplySoundStateOptions(AudioSource soundInstance, SoundStateOptions options, Guid id)
+        {
+            if (options != null)
+            {
+                //pause must happen before other sound state changes
+                if (options.paused != null && options.paused.Value == true)
+                {
+                    var index = _unpausedSoundInstances.FindIndex(x => x == id);
+                    if (index >= 0)
+                    {
+                        _unpausedSoundInstances.RemoveAt(index);
+                        soundInstance.Pause();
+                    }
+                }
+
+                if (options.Volume != null)
+                {
+                    soundInstance.volume = options.Volume.Value;
+                }
+                if (options.Pitch != null)
+                {
+                    //convert from halftone offset (-12/0/12/24/36) to pitch multiplier (0.5/1/2/4/8).
+                    soundInstance.pitch = Mathf.Pow(2.0f, (options.Pitch.Value / 12.0f));
+                }
+                if (options.Looping != null)
+                {
+                    soundInstance.loop = options.Looping.Value;
+                }
+                if (options.Doppler != null)
+                {
+                    soundInstance.dopplerLevel = options.Doppler.Value;
+                }
+                if (options.MultiChannelSpread != null)
+                {
+                    soundInstance.spread = options.MultiChannelSpread.Value * 180.0f;
+                }
+                if (options.RolloffStartDistance != null)
+                {
+                    soundInstance.minDistance = options.RolloffStartDistance.Value;
+                    soundInstance.maxDistance = options.RolloffStartDistance.Value * 1000000.0f;
+                }
+
+                //unpause must happen after other sound state changes
+                if (options.paused != null && options.paused.Value == false)
+                {
+                    var index = _unpausedSoundInstances.FindIndex(x => x == id);
+                    if (index < 0)
+                    {
+                        soundInstance.UnPause();
+                        _unpausedSoundInstances.Add(id);
+                    }
+                }
+
+
+            }
+        }
+
+        public void Update()
+        {
+            //garbage collect expired sounds, one per frame
+            if (_soundStoppedCheckIndex >= _unpausedSoundInstances.Count)
+            {
+                _soundStoppedCheckIndex = 0;
+            }
+            else
+            {
+                var id = _unpausedSoundInstances[_soundStoppedCheckIndex];
+                AudioSource soundInstance;
+                if (_soundInstances.TryGetValue(id, out soundInstance) && !soundInstance.isPlaying)
+                {
```

Comment:
Use a scoped variable here for soundInstance as `out AudioSource soundInstance`

Reply:
Not my code, but sure. Was copy/pasted from App.

---
In reply to: [263092709](https://github.com/Microsoft/mixed-reality-extension-unity/pull/45#discussion_r263092709) [](ancestors = 263092709)